### PR TITLE
Blog App: validations and model specs

### PR DESCRIPTION
**Pull request by ZuhRa-HashiMi**
Use validations for models.
Write unit tests for models.

### CODE REVIEW

File: spec/models/comment_spec.rb [hunk 1]
```diff
@@ -1,5 +1,18 @@
...
+
+  describe 'update_post_comments_counter' do
+    let!(:user) { User.create!(name: 'John Doe', email: 'jhone@doe.com', password: '123456', post_counter: 0) }
+    let!(:post) { Post.create!(title: 'My First Post', author: user, comments_counter: 0, likes_counter: 0) }
+    let!(:comment) { Comment.new(author: user, post:) }
+
+    it 'increments the post comment counter after saving' do
+      expect { comment.save }.to change { post.reload.comments_counter }.by(1)
+    end
+  end
```
Comment (jerryowusu) [inline spec/models/comment_spec.rb]: The tests are not passing as shown below. Kindly ensure that all the test are passing before resubmitting for review. You can consider adding screenshots of the test in the pull request as evidence. :+1: 

Screenshot from 2023-03-18 15-27-55

File: spec/models/like_spec.rb [hunk 2]
```diff
@@ -1,5 +1,18 @@
...
+  describe 'update_post_likes_counter' do
+    let!(:user) { User.create!(name: 'John Doe', email: 'jhone@doe.com', password: '123456', post_counter: 0) }
+    let!(:post) { Post.create!(title: 'My First Post', author: user, comments_counter: 0, likes_counter: 0) }
+    let!(:like) { Like.new(author: user, post:) }
+
+    it 'increments the post likes counter after saving' do
+      expect { like.save }.to change { post.reload.likes_counter }.by(1)
+    end
+  end
 end
```
Comment (jerryowusu) [inline spec/models/like_spec.rb]: The tests are not passing as shown below. Kindly ensure that all the test are passing before resubmitting for review. You can consider adding screenshots of the test in the pull request as evidence. :+1: 

Screenshot from 2023-03-18 15-27-55

File: spec/models/post_spec.rb [hunk 3]
```diff
@@ -1,5 +1,44 @@
...
+    it 'limits the number of most recent comments upto 5 for the post' do
+      expect(post.recent_comments(5)).to eq([comment6, comment5, comment4, comment3, comment2])
+    end
+  end
+
+  describe 'update_posts_counter' do
+    let!(:post) { Post.create!(title: 'My First Post', author: user, comments_counter: 0, likes_counter: 0) }
+
+    it 'increments the author post counter after saving' do
+      expect { post.save }.to change { user.reload.post_counter }.by(1)
```
Comment (jerryowusu) [inline spec/models/post_spec.rb:41]: The tests are not passing as shown below. Kindly ensure that all the test are passing before resubmitting for review. You can consider adding screenshots of the test in the pull request as evidence. :+1: 

Screenshot from 2023-03-18 15-27-55

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -1,5 +1,44 @@
...
+                                 created_at: 2.days.ago)
+
+      expect(user.recent_posts(2)).to eq([post1, post2])
+      expect(user.recent_posts(1)).to eq([post1])
+    end
+
+    it 'returns an empty array if the user has no posts' do
+      expect(user.recent_posts(3)).to eq([])
+    end
+  end
```
Comment (jerryowusu) [inline spec/models/user_spec.rb]: The tests are not passing as shown below. Kindly ensure that all the test are passing before resubmitting for review. You can consider adding screenshots of the test in the pull request as evidence. :+1: 

Screenshot from 2023-03-18 15-27-55

**Review by jerryowusu — CHANGES_REQUESTED** [review]
Hi @ZuhRa-HashiMi , 

Good job so far!
Your project is looking fantastic so far :star_struck: 

However, there are few changes you need to make in order to move on to the next level :sunglasses:

almost there

## Highlights
Great effort so far :heavy_check_mark: 
No linter errors :heavy_check_mark: 

## Required Changes ♻️

- [ ] Kindly add the following validations to the User model since they are part of the project requirements: 

        1.  Name must not be blank
        2.  PostsCounter must be an integer greater than or equal to zero.

- [ ]  Kindly add the following validations to the User model since they are part of the project requirements: 
       
         1.  Title must not be blank.
         2. Title must not exceed 250 characters.
         3.  CommentsCounter must be an integer greater than or equal to zero.
         4.  LikesCounter must be an integer greater than or equal to zero.


_Check the comments under the review for additional reviews._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by tufoinnkuo10 — CHANGES_REQUESTED** [review]
Hi @ZuhRa-HashiMi 

There are some issues that you still need to work on to go to the next project but you are almost there!

To Highlight
✅ You used the correct Gitflow
✅ Descriptive pull request message
✅ linter error checks validated
✅ Documented your work professionally


Required Changes ♻️
- [ ] please like the previous reviewer noted, kindly add validation checks for your user and post models.
- [ ] please your unit tests are not passing. as the previous reviewer noted, you can consider adding screenshots if fixed and they are passing on your end. please also check if you have pushed your code as your changes ain't showing yet.

tufoin

Optional suggestions
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise.
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using this [form.](https://airtable.com/shrQAqnBwek5a0O0s)

**Comment by ZuhRa-HashiMi** [comment]
# All test are passed, you can see the below picture.
test

**Review by CollinsTatang — APPROVED** [review]
# Approved 👏  
### **Highlights 🚀 :heavy_check_mark: :heavy_check_mark:** 

- The repo looks professional.:heavy_check_mark:
- Git/GitHub flow best practices have been followed:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
### **Congratulations!** 🎉 
image



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @CollinsTatang in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**